### PR TITLE
Adding swagger as a local entry in the APIM Gateway

### PR DESCRIPTION
**Pull request by ChamodDamitha**

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+     * @throws AxisFault
+     */
+
+    public Boolean addLocalEntry(String content) throws AxisFault {
+
+        Boolean value;
+        try {
+            value = localEntryAdminServiceStub.addEntry(content);
+        } catch (RemoteException e) {
+            throw new AxisFault("Error occurred while generating the response ", e.getMessage(), e);
```
Comment (hasuniea) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java]: passing throwable object is enough.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+     * @param key API Id to be retrived
+     * @return LocalEntry for the given API
+     * @throws AxisFault
+     */
+    public Object getEntry(String key) throws AxisFault {
+        Object object;
+        try {
+            object = localEntryAdminServiceStub.getEntry(key);
+        } catch (RemoteException e) {
+            throw new AxisFault("Error occurred while retrieving the local entry", e.getMessage(), e);
```
Comment (hasuniea) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java]: Here as well. No need to pass e.getMessage with the throwable object.

File: service-stubs/apimgt/org.wso2.carbon.apimgt.localentry.stub/pom.xml [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>org.wso2.carbon.apimgt.localentry.stub</artifactId>
+    <packaging>bundle</packaging>
+    <name>WSO2 Carbon - Locale Entry Stub</name>
+    <url>http://wso2.org</url>
+
+    <build>
+        <plugins>
+
```
Comment (hasuniea) [inline service-stubs/apimgt/org.wso2.carbon.apimgt.localentry.stub/pom.xml]: remove the new line.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/LocalEntryAdminClient.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            authAdminStub.login(userName, password, host);
+            ServiceContext serviceContext = authAdminStub.
+                    _getServiceClient().getLastOperationContext().getServiceContext();
+            return (String) serviceContext.getProperty(HTTPConstants.COOKIE_STRING);
+        } catch (RemoteException e) {
+            throw new AxisFault("Error while contacting the authentication admin services", e);
+        } catch (LoginAuthenticationExceptionException e) {
+            throw new AxisFault("Error while authenticating against the API gateway admin", e);
+        }
+
```
Comment (kavishkafernando) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/LocalEntryAdminClient.java]: Shall we remove this extra line?

**Comment by hasuniea** [comment]
Can you please check the integration tests as well. It might fail with this change because by default this feature is enabled.

File: service-stubs/apimgt/org.wso2.carbon.apimgt.localentry.stub/pom.xml [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  ~  See the License for the specific language governing permissions and
+  ~  limitations under the License.
+  -->
+
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+    <parent>
+        <groupId>org.wso2.carbon.apimgt</groupId>
+        <artifactId>apimgt-stubs</artifactId>
+        <version>6.4.155-SNAPSHOT</version>
```
Comment (hasuniea) [inline service-stubs/apimgt/org.wso2.carbon.apimgt.localentry.stub/pom.xml]: the version should be changed. Update it to 6.4.163-SNAPSHOT

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/service/APILocalEntryAdmin.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
+package org.wso2.carbon.apimgt.gateway.service;
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/service/APILocalEntryAdmin.java]: Shall we add licence headers?

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
+package org.wso2.carbon.apimgt.gateway.utils;
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java]: Missing licence headers

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    /**
+     * Add Local entry
+     *
+     * @param content Swagger Content
+     * @return Status of the add operation
+     * @throws AxisFault
+     */
+
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java]: Lets remove this extra line

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        Boolean value;
+        try {
+            value = localEntryAdminServiceStub.addEntry(content);
+        } catch (RemoteException e) {
+            throw new AxisFault("Error occurred while generating the response ", e.getMessage(), e);
+        } catch (LocalEntryAdminException e) {
+            throw new AxisFault("Error occurred while adding the local entry", e.getMessage(), e);
+        }
+        return value;
+
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java]: Extra line

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    /**
+     * Delete the local entry
+     *
+     * @param key APT Id to be deleted
+     * @return Stataus of the delete operation
+     * @throws AxisFault
+     */
+
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java]: Extra line

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java [hunk 11]
```diff
@@ -579,6 +579,7 @@ public static boolean jsFunction_manageAPI(Context cx, Scriptable thisObj, Objec
 
             //Save swagger in the registry
             apiProvider.saveSwagger20Definition(api.getId(), (String) apiData.get("swagger", apiData));
+            apiProvider.addSwaggerToLocalEntry(api, (String) apiData.get("swagger", apiData));
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java]: Shall we move "swagger" to a string constant?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 12]
```diff
@@ -208,6 +209,55 @@ protected String getUserNameWithoutChange() {
         return userNameWithoutChange;
     }
 
+    @Override
+    public void addSwaggerToLocalEntry(API api, String jsonText) {
+        if (log.isDebugEnabled()) {
+            log.debug("Adding a new Local Entry");
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: Lets make this debug log more meaningful by adding the api too.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 13]
```diff
@@ -208,6 +209,55 @@ protected String getUserNameWithoutChange() {
...
+            Environment environment = environments.get(environmentName);
+            api.getEnvironments();
+            try {
+                localEntryAdminClient = new LocalEntryAdminClient(api.getId(), environment);
+                localEntryAdminClient.deleteEntry(api.getId().toString());
+                localEntryAdminClient.addLocalEntry("<localEntry key=\"" + api.getId() + "\">" +
+                        jsonText.replaceAll("&(?!amp;)", "&amp;").
+                                replaceAll("<","&lt;").replaceAll(">","&gt;") + "</localEntry>");
+            } catch (AxisFault e) {
+                log.error("Error occurred while Deleting the local entry ", e);
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: Lets add the api name here too

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/LocalEntryAdminClient.java [hunk 14]
```diff
@@ -0,0 +1,120 @@
+package org.wso2.carbon.apimgt.impl.utils;
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/LocalEntryAdminClient.java]: Missing licence headers

File: service-stubs/apimgt/org.wso2.carbon.apimgt.localentry.stub/pom.xml [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+                                </source>
+                            </sources>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+
```
Comment (npamudika) [inline service-stubs/apimgt/org.wso2.carbon.apimgt.localentry.stub/pom.xml]: Lets remove this extra line

File: service-stubs/apimgt/pom.xml [hunk 16]
```diff
@@ -33,6 +33,7 @@
         <module>org.wso2.carbon.apimgt.gateway.stub</module>
         <module>org.wso2.carbon.apimgt.statsupdate.stub</module>
         <module>org.wso2.carbon.apimgt.tier.cache.stub</module>
-	<module>org.wso2.carbon.apimgt.registry.cache.stub</module>
+	    <module>org.wso2.carbon.apimgt.registry.cache.stub</module>
```
Comment (npamudika) [inline service-stubs/apimgt/pom.xml]: Formatting error

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java (hunk 1 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java]: Updated

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java (hunk 2 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java]: Updated

File: service-stubs/apimgt/org.wso2.carbon.apimgt.localentry.stub/pom.xml (hunk 3 above)
Comment (ChamodDamitha) [reply service-stubs/apimgt/org.wso2.carbon.apimgt.localentry.stub/pom.xml]: Updated

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/LocalEntryAdminClient.java (hunk 4 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/LocalEntryAdminClient.java]: Updated

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/service/APILocalEntryAdmin.java (hunk 6 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/service/APILocalEntryAdmin.java]: Updated

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java (hunk 7 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java]: Updated

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java (hunk 8 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java]: Updated

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java (hunk 9 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java]: Updated

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java (hunk 10 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/LocalEntryClient.java]: Updated

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java (hunk 11 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java]: Updated

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 12 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: Updated

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 13 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: Updated

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/LocalEntryAdminClient.java (hunk 14 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/LocalEntryAdminClient.java]: Updated

File: service-stubs/apimgt/org.wso2.carbon.apimgt.localentry.stub/pom.xml (hunk 15 above)
Comment (ChamodDamitha) [reply service-stubs/apimgt/org.wso2.carbon.apimgt.localentry.stub/pom.xml]: Updated

File: service-stubs/apimgt/pom.xml (hunk 16 above)
Comment (ChamodDamitha) [reply service-stubs/apimgt/pom.xml]: Updated

File: service-stubs/apimgt/org.wso2.carbon.apimgt.localentry.stub/pom.xml (hunk 5 above)
Comment (ChamodDamitha) [reply service-stubs/apimgt/org.wso2.carbon.apimgt.localentry.stub/pom.xml]: Updated

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/service/APILocalEntryAdmin.java (hunk 6 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/service/APILocalEntryAdmin.java]: Updated

**Pull request merged**